### PR TITLE
[indexer grpc] compression support.

### PR DESCRIPTION
**Pull request by larry-aptos**
### Description

* Compression support for indexer grpc stack.
  * The feature is optional and backward compatible; in other words, community should not expect any surprise with this change. 
* TODO: better documents once we're ready to ship this to community. 

### Test Plan
* Added unit tests for the compression component. 
* Tested with production traffic and looks good overall.

### CODE REVIEW

File: ecosystem/indexer-grpc/indexer-grpc-cache-worker/src/worker.rs [hunk 1]
```diff
@@ -205,90 +211,74 @@ async fn process_transactions_from_node_response(
...
+                let last_transaction_pb_timestamp = last_transaction.timestamp.clone();
+
+                log_grpc_step(
+                    SERVICE_TYPE,
+                    IndexerGrpcStep::CacheWorkerReceivedTxns,
+                    Some(start_version as i64),
+                    Some(last_transaction_version as i64),
+                    first_transaction_pb_timestamp.as_ref(),
+                    last_transaction_pb_timestamp.as_ref(),
+                    Some(batch_start_time.elapsed().as_secs_f64()),
```
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-grpc-cache-worker/src/worker.rs:237]: `batch_start_time` gets reset after each batch. can we reset it before each download? we want to measure the ony the download duration here

File: ecosystem/indexer-grpc/indexer-grpc-cache-worker/src/lib.rs [hunk 2]
```diff
@@ -17,18 +17,21 @@ pub struct IndexerGrpcCacheWorkerConfig {
     pub fullnode_grpc_address: Url,
     pub file_store_config: IndexerGrpcFileStoreConfig,
     pub redis_main_instance_address: RedisUrl,
+    pub enable_cache_compression: bool,
```
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-grpc-cache-worker/src/lib.rs:20]: set a default value

File: ecosystem/indexer-grpc/indexer-grpc-data-service/src/service.rs [hunk 3]
```diff
@@ -520,95 +466,87 @@ async fn data_fetch(
...
-                start_version_timestamp.as_ref(),
-                end_version_timestamp.as_ref(),
-                Some(cache_fetch_start_time.elapsed().as_secs_f64()),
+                start_version_timestamp,
+                end_version_timestamp,
+                Some(duration_in_secs),
+                Some(size_in_bytes),
+                Some(num_of_transactions as i64),
+                Some(request_metadata.clone()),
+            );
```
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-grpc-data-service/src/service.rs:497]: this should be moved to before decoding starts

File: ecosystem/indexer-grpc/indexer-grpc-file-store/src/lib.rs [hunk 4]
```diff
@@ -17,6 +17,7 @@ pub struct IndexerGrpcFileStoreWorkerConfig {
     pub redis_main_instance_address: RedisUrl,
     pub enable_expensive_logging: Option<bool>,
     pub chain_id: u64,
+    pub enable_cache_compression: bool,
```
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-grpc-file-store/src/lib.rs:20]: set default value

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/cache_operator.rs [hunk 5]
```diff
@@ -135,6 +147,21 @@ impl<T: redis::aio::ConnectionLike + Send + Clone> CacheOperator<T> {
         self.get_config_by_key(CACHE_KEY_LATEST_VERSION).await
     }
 
+    /// Returns starting version and ending version.
+    pub async fn get_latest_starting_and_ending_verisons(
```
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-grpc-utils/src/cache_operator.rs:151]: is this used anywhere?

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/cache_operator.rs [hunk 6]
```diff
@@ -186,34 +213,101 @@ impl<T: redis::aio::ConnectionLike + Send + Clone> CacheOperator<T> {
...
             Ok(CacheCoverageStatus::CacheHit(std::cmp::min(
                 latest_version - requested_version,
-                BLOB_STORAGE_SIZE as u64,
+                FILE_ENTRY_TRANSACTION_COUNT,
             )))
         }
     }
 
+    /// Fail if not all transactions requested are returned
+    pub async fn batch_get_encoded_proto_data_x(
```
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-grpc-utils/src/cache_operator.rs:222]: is this used anywhere?

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/cache_operator.rs [hunk 7]
```diff
@@ -252,54 +366,57 @@ impl<T: redis::aio::ConnectionLike + Send + Clone> CacheOperator<T> {
         }
     }
 
-    // TODO: Remove this
-    pub async fn batch_get_encoded_proto_data(
+    pub async fn get_transactions_with_durations(
```
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-grpc-utils/src/cache_operator.rs:369]: this isn't used anywhere? if it's for refactoring we should split it out into another PR

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/config.rs [hunk 8]
```diff
@@ -10,11 +10,13 @@ pub struct GcsFileStore {
     pub gcs_file_store_bucket_name: String,
     // Required to operate on GCS.
     pub gcs_file_store_service_account_key_path: String,
+    pub enable_compression: bool,
```
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-grpc-utils/src/config.rs:13]: set default value

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/file_store_operator/gcs.rs [hunk 9]
```diff
@@ -75,14 +87,31 @@ impl FileStoreOperator for GcsFileStoreOperator {
...
 
-    /// Gets the raw transactions file from the file store. Mainly for verification purpose.
-    async fn get_raw_transactions(&self, version: u64) -> anyhow::Result<TransactionsFile> {
-        let batch_start_version = version / BLOB_STORAGE_SIZE as u64 * BLOB_STORAGE_SIZE as u64;
-        let current_file_name = generate_blob_name(batch_start_version);
-        let bytes = Object::download(&self.bucket_name, current_file_name.as_str()).await?;
-        serde_json::from_slice(&bytes)
-            .map_err(|e| anyhow::anyhow!("Failed to deserialize transactions file: {}", e))
-    }
+    // /// Gets the transactions files from the file store. version has to be a multiple of BLOB_STORAGE_SIZE.
```
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-grpc-utils/src/file_store_operator/gcs.rs:90]: delete this

**Review by rtso — CHANGES_REQUESTED** [review]
can we also test this code to see if it works with the old uncompressed formats

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/cache_operator.rs (hunk 6 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-grpc-utils/src/cache_operator.rs]: yea; rename it to something meaningful. we'll address it with refactoring.

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/file_store_operator/gcs.rs [hunk 10]
```diff
@@ -155,27 +162,40 @@ impl FileStoreOperator for GcsFileStoreOperator {
...
         );
-
+        let start_time = std::time::Instant::now();
         let bucket_name = self.bucket_name.clone();
-        let current_batch = transactions.iter().cloned().collect_vec();
-        let transactions_file = build_transactions_file(current_batch).unwrap();
+        let file_entry: FileEntry =
+            FileEntryBuilder::new(transactions, self.storage_format).try_into()?;
+        let file_entry_key = FileEntryKey::new(start_version, self.storage_format).to_string();
+        log_grpc_step(
```
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-grpc-utils/src/file_store_operator/gcs.rs:183]: this is redundant, this step is already logged in the filestore worker. we do need to log the filestore compression step though.

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/cache_operator.rs (hunk 5 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-grpc-utils/src/cache_operator.rs:151]: yea

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/cache_operator.rs (hunk 7 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-grpc-utils/src/cache_operator.rs:369]: yes, but this collides with the function `batch_get_encoded_proto_data_x`; just rename here to avoid further confusion; we can address in refactoring.

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/config.rs (hunk 8 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-grpc-utils/src/config.rs:14]: done.

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs [hunk 11]
```diff
@@ -0,0 +1,628 @@
...
+    type Error = anyhow::Error;
+
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        let metadata = serde_json::from_slice(value.as_slice())
+            .context("FileStoreMetadata json deserialization failed.")?;
+        Ok(metadata)
+    }
+}
+
+pub struct CacheEntryBuilder {
```
[the remark this answers is not in the document]
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs:113]: makes sense ... the redundant struct was built to upcast to a specific format from bytes.

[the remark this answers is not in the document]
Comment (rtso) [reply ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs:113]: yeah i was reading about this TryInto / TryFrom pattern. it's used for type casting, which is a bit different from how we need to use it here. i think having a single struct for CacheEntry and FileEntry that takes care of 1) building the key, 2) compressing Transaction(s) into entry, 3) decompressing entry to Transaction(s) would be much simpler and cleaner. let me know what you think.

[the remark this answers is not in the document]
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs]: yea makes sense. done

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs [hunk 12]
```diff
@@ -0,0 +1,513 @@
...
+    pub starting_version: u64,
+    /// The transactions in the blob.
+    #[serde(rename = "transactions")]
+    pub transactions_in_base64: Vec<String>,
+}
+
+/// FileStoreMetadata is the metadata for the file store.
+/// It's a JSON file with name: metadata.json.
+#[derive(Serialize, Deserialize, Copy, Clone, Debug, PartialEq, Eq)]
+pub struct FileStoreMetadata {
```
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs:37]: we can also use the same single struct pattern for FileStoreMetadata now

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs [hunk 13]
```diff
@@ -0,0 +1,513 @@
...
+                panic!("JsonBase64UncompressedProto is not supported in cache.")
+            },
+        }
+    }
+}
+
+impl TryFrom<CacheEntry> for Transaction {
+    type Error = anyhow::Error;
+
+    fn try_from(value: CacheEntry) -> Result<Self, Self::Error> {
```
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs:155]: could we move this into a function like `CacheEntry::to_transaction`?

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs [hunk 14]
```diff
@@ -0,0 +1,513 @@
...
+                panic!("Base64UncompressedProto is not supported.")
+            },
+        }
+    }
+}
+
+impl TryFrom<FileEntry> for TransactionsInStorage {
+    type Error = anyhow::Error;
+
+    fn try_from(value: FileEntry) -> Result<Self, Self::Error> {
```
Comment (rtso) [inline ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs:286]: same thing here. could we move this into a function `FileEntry::to_transactions`

**Review by rtso — COMMENTED** [review]
looks a lot easier to understand now, thanks for making the changes. left some more comments :)

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs (hunk 12 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs:36]: done.

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs (hunk 13 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs]: done.

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs (hunk 14 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs]: done.

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/file_store_operator/gcs.rs (hunk 10 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-grpc-utils/src/file_store_operator/gcs.rs:179]: done.

**Comment by larry-aptos** [comment]
> can we also test this code to see if it works with the old uncompressed formats

done.

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/file_store_operator/gcs.rs [hunk 15]
```diff
@@ -137,7 +137,11 @@ impl FileStoreOperator for GcsFileStoreOperator {
         chain_id: u64,
         version: u64,
     ) -> anyhow::Result<()> {
-        let metadata = FileStoreMetadata::new(chain_id, version);
+        let metadata = FileStoreMetadata::new(
+            chain_id,
+            version,
+            StorageFormat::JsonBase64UncompressedProto,
```
Comment (bowenyang007) [inline ecosystem/indexer-grpc/indexer-grpc-utils/src/file_store_operator/gcs.rs:143]: might not want to hardcode this.

Comment (rtso) [reply ecosystem/indexer-grpc/indexer-grpc-utils/src/file_store_operator/gcs.rs:143]: hmm does FileStoreMetadata even need to specify a storage format? it should be unaffected by this compression change. if we don't need storage format, we should remove it as a var.

**Review by rtso — APPROVED** [review]

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/file_store_operator/gcs.rs (hunk 15 above)
Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-grpc-utils/src/file_store_operator/gcs.rs]: added the check; this prevents people trying to mix compression with uncompressed stuff.

File: ecosystem/indexer-grpc/indexer-grpc-utils/src/file_store_operator/local.rs [hunk 16]
```diff
@@ -34,20 +47,16 @@ impl FileStoreOperator for LocalFileStoreOperator {
...
 
-    async fn get_transactions(&self, version: u64) -> anyhow::Result<Vec<String>> {
-        let batch_start_version = version / BLOB_STORAGE_SIZE as u64 * BLOB_STORAGE_SIZE as u64;
-        let current_file_name = generate_blob_name(batch_start_version);
-        let file_path = self.path.join(current_file_name);
+    fn storage_format(&self) -> StorageFormat {
+        self.storage_format
+    }
+
+    async fn get_transactions_bytes(&self, version: u64) -> anyhow::Result<Vec<u8>> {
```
Comment (bowenyang007) [inline ecosystem/indexer-grpc/indexer-grpc-utils/src/file_store_operator/local.rs:54]: Can you name this get_transactions_old or uncompressed or something?

Comment (larry-aptos) [reply ecosystem/indexer-grpc/indexer-grpc-utils/src/file_store_operator/local.rs]: done.

**Review by bowenyang007 — APPROVED** [review]
